File: lib/rubyfatt2/customer.ex
defmodule Rubyfatt2.Customer do
  use Ecto.Schema
  import Ecto.Changeset

  schema "customers" do
    belongs_to :user, Rubyfatt2.User
    has_many :slips, Rubyfatt2.Slip
    has_many :invoice_projects, Rubyfatt2.InvoiceProject
    has_many :invoices, Rubyfatt2.Invoice

    field :title, :string
    field :name, :string
    field :surname, :string
    field :address, :string
    field :zip_code, :string
    field :town, :string
    field :province, :string
    field :country, :string
    field :tax_code, :string
    field :vat, :string
    field :info, :string

    timestamps()
  end

  def changeset(customer, attrs) do
    customer
    |> cast(attrs, [
      :user_id,
      :title,
      :name,
      :surname,
      :address,
      :zip_code,
      :town,
      :province,
      :country,
      :tax_code,
      :vat,
      :info
    ])
    |> validate_required([:title])
    |> unique_constraint(:title, name: :customers_user_id_title_index)
  end

  def full_info(c) do
    "#{c.title}\n#{c.address}\n#{c.zip_code} #{c.town}\nC.F. #{c.tax_code}\nP.IVA #{c.vat}"
  end
end

File: frontend/src/models/customer.ts
export interface ICustomer {
    id?: number;
    title: string;
    name: string;
    surname: string;
    address: string;
    zip_code: string;
    town: string;
    province: string;
    country: string;
    tax_code: string;
    vat: string;
    info: string;
}

export default class Customer {
    public id: number;
    public title: string;
    public name: string;
    public surname: string;
    public address: string;
    public zip_code: string;
    public town: string;
    public province: string;
    public country: string;
    public tax_code: string;
    public vat: string;
    public info: string;

    constructor(c: ICustomer) {
        this.id = c.id;
        this.title = c.title;
        this.name = c.name;
        this.surname = c.surname;
        this.address = c.address;
        this.zip_code = c.zip_code;
        this.town = c.town;
        this.province = c.province;
        this.country = c.country;
        this.tax_code = c.tax_code;
        this.vat = c.vat;
        this.info = c.info;
    }

    public get fullName(): string {
        return `${this.name || ""} ${this.surname || ""}`;
    }

    public get fullAddress(): string {
        return `${this.address || ""} ${this.zip_code || ""} ${this.town || ""}
                ${this.province || ""} ${this.country || ""}`;
    }
}
